diffkemp/simpll/InstPatternComparator.h

```diff
@@ -0,0 +1,125 @@
+//===----- InstPatternComparator.h - Code pattern instruction matcher -----===//
+//
+//       SimpLL - Program simplifier for analysis of semantic difference      //
+//
+// This file is published under Apache 2.0 license. See LICENSE for details.
+// Author: Petr Silling, psilling@redhat.com
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file contains the declaration of the LLVM code pattern matcher. The
+/// pattern matcher is a comparator extension of the LLVM FunctionComparator
+/// tailored to comparison of general instruction-based patterns.
+///
+//===----------------------------------------------------------------------===//
+
+#ifndef DIFFKEMP_SIMPLL_INSTPATTERNCOMPARATOR_H
+#define DIFFKEMP_SIMPLL_INSTPATTERNCOMPARATOR_H
+
+#include "FunctionComparator.h"
+#include "PatternSet.h"
+#include <unordered_map>
+#include <vector>
+
+using namespace llvm;
+
+/// Extension of LLVM FunctionComparator which compares a difference pattern
+/// against its corresponding module function. Compared functions are expected
+/// to lie in different modules. The module function is expected on the left
+/// side while the pattern function is expected on the right side.
+class InstPatternComparator : protected FunctionComparator {
+  public:
+    /// Pattern instructions matched to their respective module replacement
+    /// instructions. Pattern instructions are used as keys.
+    mutable InstructionMap InstMatchMap;
+    /// Pattern input instructions and arguments matched to module input
+    /// instructions and arguments. Pattern input is used for keys.
+    mutable InputMap InputMatchMap;
+    /// Module input instructions and arguments matched to pattern input
+    /// instructions and arguments. Module input is used for keys.
+    mutable InputMap ReverseInputMatchMap;
```

Comment:
This should be possible. However, I believe it would reduce code clarity.

Reply:
Well, since they're *always* updated at the same time, some kind of bi-directional map would make more sense here, but I'm afraid that neither LLVM nor std has such a thing.

Could you at least change the names? It's not very clear which is which now.

Reply:
I agree with both. For now then, I will be changing the names to `PatMatchMap` and `ModMatchMap`. This should more clearly say which side is used for keys.